rust/src/shards/ws.rs

```diff
@@ -0,0 +1,323 @@
+use crate::{
+  core::{run_blocking, registerShard, BlockingShard},
+  types::{
+    ClonedVar, Context, ExposedInfo, ExposedTypes, ParamVar, Parameters, Type, Types, Var,
+    ANYS_TYPES, FRAG_CC, NONE_TYPES, STRING_TYPES,
+  },
+  Shard,
+};
+use std::net::TcpStream;
+use std::rc::Rc;
+use std::thread::spawn;
+use tungstenite::{
+  accept, handshake::client::Response, stream::MaybeTlsStream, Message, WebSocket,
+};
+
+static WS_CLIENT_TYPE: Type = Type::object(FRAG_CC, 2004042604); // 'wsCl'
+static WS_CLIENT_SLICE: &'static [Type] = &[WS_CLIENT_TYPE];
+static WS_CLIENT_VAR: Type = Type::context_variable(WS_CLIENT_SLICE);
+
+lazy_static! {
+  static ref WS_CLIENT_VEC: Types = vec![WS_CLIENT_TYPE];
+  static ref WS_CLIENT_VAR_VEC: Types = vec![WS_CLIENT_VAR];
+  static ref WS_CLIENT_USER_PARAMS: Parameters = vec![(
+    cstr!("Client"),
+    cstr!("The WebSocket client instance."),
+    &WS_CLIENT_VAR_VEC[..],
+  )
+    .into(),];
+}
+
+type WsClient = WebSocket<MaybeTlsStream<TcpStream>>;
+
+#[derive(Default)]
+struct Client {
+  ws: Rc<Option<WsClient>>,
+}
+
+impl Shard for Client {
+  fn registerName() -> &'static str
+  where
+    Self: Sized,
+  {
+    cstr!("WS.Client")
+  }
+
+  fn hash() -> u32
+  where
+    Self: Sized,
+  {
+    compile_time_crc32::crc32!("WS.Client-rust-0x20200101")
+  }
+
+  fn name(&mut self) -> &str {
+    "WS.Client"
+  }
+
+  fn inputTypes(&mut self) -> &Types {
+    &STRING_TYPES
+  }
+
+  fn outputTypes(&mut self) -> &Types {
+    &WS_CLIENT_VEC
+  }
+
+  fn activate(&mut self, context: &Context, input: &Var) -> Result<Var, &str> {
+    Ok(run_blocking(self, context, input))
+  }
+}
+
+impl BlockingShard for Client {
+  fn activate_blocking(&mut self, _context: &Context, input: &Var) -> Result<Var, &str> {
+    let addr: &str = input.try_into()?;
+    let ws = tungstenite::client::connect(addr).map_err(|e| {
+      shlog!("{}", e);
+      "Failed to connect to server."
+    })?;
+    // we need to keep this alive here, otherwise it will be dropped at the end of this function
+    self.ws = Rc::new(Some(ws.0));
+    let var_ws = Var::new_object(&self.ws, &WS_CLIENT_TYPE);
+    Ok(var_ws)
+  }
+}
+
+#[derive(Default)]
+struct ClientUser {
+  ws: Rc<Option<WsClient>>,
```

Comment:
Why not `Option<Rc<WsClient>>` or even just `Option<WsClient>`? It doesn't need to stay alive past the lifetime of the `Client` shard I believe.

Reply:
yeah has to be that way so the client could even disappear